### DISCUSSION TITLE
Optional implementation of `Source::skip(byteCount: Long): Long`

### DISCUSSION THREAD

**Discussion by Nava2** [square/okio#1330] [Ideas]
I've got a use case where I need to implement skipping efficiently due to a file type. I wonder if it would be feasible to add a signature to `Source` like:

```kotlin
  /**
   * Iterates the internal source forward by up-to `byteCount`.
   * @return Number of bytes skipped, -1 if the operation failed (e.g. exhausted).
   */
  fun skip(byteCount: Long): Long = Buffer().use { read(it, byteCount) }
```

Some bias, I'm trying to avoid reimplementing `BufferedSource`'s `skip(...)` to monkey-patch an implementation to cast down to a separate interface.

An alternative would be to use a different interface and do a `is SkippableSource` or similar when calling `Buffer.skip()`. 

I'm curious if this is anything taht would be considered.

**Comment by swankjesse** [comment]
I think the function you’ve written is a good solution to this problem. It’s potentially easier to write code like this vs. built-ins in Okio.

As for optimized skipping… I’m a bit skeptical that it’s worth the effort. You can use FileHandle to random access a file, but skipping on a stream is making an inefficient path faster. Generally better to pick data structures or APIs where skipping is either efficient or unnecessary.

**Reply by Nava2** [reply]
I agree with skipping in a stream being inefficient generally. For my use case (bias!), the underlying sources are typically setup to be pointer based and as such easily allow random pointer manipulation. To simplify usage/implementation I'm looking to avoid random seeking and only do forward skipping.

I really like okio's patterns and abstractions, but maybe I'm trying too hard to make an abstraction fit that doesn't. A separately defined API work some converters is probably the right call for me if it's not something y'all are interested in. 

I just realized I can probably get a _reasonable_ implementation of a "Buffered source" without monkey patching by wrapping BufferedInputStream.

Just to clarify, which code were you referring to, Jesse?

**Reply by swankjesse** [reply]
The heap reader in Shark is a good example of layering Okio on a random access file: https://github.com/square/leakcanary/blob/main/shark/shark-hprof/src/main/java/shark/RandomAccessHprofReader.kt
